Pkg 1550 1.17.0 plus linux cuda

danpetry opened the pull request:
**Destination channel:** defaults

### Links

- [PKG-1550](https://anaconda.atlassian.net/browse/PKG-1550?atlOrigin=eyJpIjoiOGI0ZTNkMWYyNmZlNDc0ZThlOTBmODk4YmM1OTY3ZTkiLCJwIjoiaiJ9) 
- [Upstream repository](https://github.com/microsoft/onnxruntime/tree/v1.17.0)
- [Upstream changelog/diff](https://github.com/microsoft/onnxruntime/releases)

### Explanation of changes:

- Update to support CUDA 11.8 for the ticket; CPU versions updated alongside
- Tests enabled and various fixes related to that
- NB that support for CentOS 7 was dropped upstream (see changelog). Tests being run verifies that it's still ok.
- CUDA version is building successfully on a build machine, log [here](https://github.com/AnacondaRecipes/onnxruntime-feedstock/files/14638594/18buildafterfinaltidy.log.zip)
- see commit messages for details for changes.

[PKG-1550]: https://anaconda.atlassian.net/browse/PKG-1550?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

danpetry committed:
Build for CUDA execution provider for linux-64

- Tell nvcc to use conda-provided compiler
- Provide CUDA variant in cbc.yaml
- Enable cuda profiling
- Add required CUDA runtime packages
- Specify cuda architecture explicitly
- Handle wheel names for CPU vs CUDA variants

danpetry committed:
Enable unit testing

- skip CUDA test that's unsupported by g3x GPU
- skip custom op test which is segfaulting on linux cuda (both C++ and python API variants)
- skip onnx dependent tests; including onnx in the host environment leads to build errors

The segfault goes away when our libstdc++ is replaced by one from the system.
Work slated to correct the issue in that library, and we'll release this one.
See issue mentioned in patch for further details.

danpetry committed:
Update version and sha to 1.17.1

danpetry committed:
Use ninja for Windows builds

Fixes required, because the default is VS2022 and adds an extra folder depth
compared with ninja:
 - Fix cwd name on windows; upstream adds the config to the pathname twice
 - Fix the xcopy path for the same reason

danpetry committed:
Adjust patches

- Remove upstreamed patch 16337.patch
- Remove patch that removes now-passing test
- Rebase other patches
- move patches to patches/ directory
- Use windows line endings for Windows patch

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_fc95n2xvag/clone` failed. (See above for error)
The following problems have been found:

===== WARNINGS =====
- clone/recipe/meta.yaml:42: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:55: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:44: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:43: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:46: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:47: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:56: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:58: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.

===== ERRORS =====
- clone/recipe/build.sh:68: pip_install_args: `pip install` should be run with --no-deps and --no-build-isolation.
===== Final Report: =====
1 Error and 8 Warnings were found

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_centbewiav/clone` failed. (See above for error)
The following problems have been found:

===== WARNINGS =====
- clone/recipe/meta.yaml:55: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:56: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:46: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:47: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:58: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:44: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:43: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:42: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.

===== ERRORS =====
- clone/recipe/build.sh:67: pip_install_args: `pip install` should be run with --no-deps and --no-build-isolation.
===== Final Report: =====
1 Error and 8 Warnings were found

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_86yk57_fze/clone` failed. (See above for error)
The following problems have been found:

===== WARNINGS =====
- clone/recipe/meta.yaml:43: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:55: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:42: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:47: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:44: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:46: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:58: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:56: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.

===== ERRORS =====
- clone/recipe/build.sh:69: pip_install_args: `pip install` should be run with --no-deps and --no-build-isolation.
===== Final Report: =====
1 Error and 8 Warnings were found

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_8fj415mhp8/clone` failed. (See above for error)
The following problems have been found:

===== WARNINGS =====
- clone/recipe/meta.yaml:47: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:42: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:44: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:43: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:56: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:46: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:58: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:55: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.

===== ERRORS =====
- clone/recipe/build.sh:69: pip_install_args: `pip install` should be run with --no-deps and --no-build-isolation.
===== Final Report: =====
1 Error and 8 Warnings were found

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_20pv8kpmsd/clone` failed. (See above for error)
The following problems have been found:

===== WARNINGS =====
- clone/recipe/meta.yaml:47: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:58: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:56: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:44: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:46: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:42: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:43: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:55: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.

===== ERRORS =====
- clone/recipe/build.sh:69: pip_install_args: `pip install` should be run with --no-deps and --no-build-isolation.
===== Final Report: =====
1 Error and 8 Warnings were found

</details>

danpetry committed:
Upgrade osx sdk to provide std::filesystem::path and std::filesystem::hash_value

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_77q2ocjhiz/clone` failed. (See above for error)
The following problems have been found:

===== WARNINGS =====
- clone/recipe/meta.yaml:48: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:43: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:45: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:47: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:58: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:44: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:56: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.

===== ERRORS =====
- clone/recipe/build.sh:69: pip_install_args: `pip install` should be run with --no-deps and --no-build-isolation.
===== Final Report: =====
1 Error and 7 Warnings were found

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_ef41lyr_o2/clone` failed. (See above for error)
The following problems have been found:

===== WARNINGS =====
- clone/recipe/meta.yaml:57: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:45: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:44: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:48: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:59: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:46: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:49: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.

===== ERRORS =====
- clone/recipe/build.sh:69: pip_install_args: `pip install` should be run with --no-deps and --no-build-isolation.
===== Final Report: =====
1 Error and 7 Warnings were found

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_9275ol_xs1/clone` failed. (See above for error)
The following problems have been found:

===== WARNINGS =====
- clone/recipe/meta.yaml:51: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:48: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:59: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:61: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:47: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:50: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:46: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.

===== ERRORS =====
- clone/recipe/build.sh:69: pip_install_args: `pip install` should be run with --no-deps and --no-build-isolation.
===== Final Report: =====
1 Error and 7 Warnings were found

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_213a2cn_fg/clone` failed. (See above for error)
The following problems have been found:

===== WARNINGS =====
- clone/recipe/meta.yaml:50: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:51: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:61: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:59: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:47: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:46: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:48: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.

===== ERRORS =====
- clone/recipe/build.sh:69: pip_install_args: `pip install` should be run with --no-deps and --no-build-isolation.
===== Final Report: =====
1 Error and 7 Warnings were found

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_6avv63ksfn/clone` failed. (See above for error)
The following problems have been found:

===== WARNINGS =====
- clone/recipe/meta.yaml:45: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:43: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:47: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:44: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:56: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:48: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:58: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.

===== ERRORS =====
- clone/recipe/build.sh:68: pip_install_args: `pip install` should be run with --no-deps and --no-build-isolation.
===== Final Report: =====
1 Error and 7 Warnings were found

</details>

danpetry committed:
Use current usual pip install args

danpetry committed:
Refine dependencies:

- libdate and optional-lite aren't needed
- zlib and pybind11 don't have any upstream requirements, nor run_exports, so we just pin to latest for reproducibility
- clarifies that gmock is a test requirement (because of this, it doesn't need to be pinned)

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_1bjkqg4vm6/clone` failed. (See above for error)
The following problems have been found:

===== WARNINGS =====
- clone/recipe/meta.yaml:53: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:43: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:45: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:56: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:54: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
===== Final Report: =====
0 Errors and 5 Warnings were found

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_2amjc1pg01/clone` failed. (See above for error)
The following problems have been found:

===== WARNINGS =====
- clone/recipe/meta.yaml:54: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:53: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:43: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:56: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:45: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
===== Final Report: =====
0 Errors and 5 Warnings were found

</details>

danpetry commented:
[CUDA build log file](https://github.com/AnacondaRecipes/onnxruntime-feedstock/files/14638594/18buildafterfinaltidy.log.zip)

JeanChristopheMorinPerso commented on recipe/meta.yaml:
```diff
@@ -39,26 +46,34 @@ requirements:
     - python-flatbuffers
     - sympy
-    - zstd
+    - {{ pin_compatible('zstd') }}
```
You could rely on the run exports here. Right now you are making the constraint more strict compared to the run_exports.

danpetry replied:
done

JeanChristopheMorinPerso commented on recipe/bld.bat:
```diff
@@ -5,17 +5,19 @@
 if errorlevel 1 exit 1
 
-xcopy /S /Y /F build-ci\Release\Release\dist\onnxruntime-*.whl onnxruntime-%PKG_VERSION%-py3-none-any.whl*
+xcopy /S /Y /F build-ci\Release\dist\onnxruntime-*.whl onnxruntime-%PKG_VERSION%-py3-none-any.whl*
```
That copy is useless (and isn't done in conda-forge). It'll slow down the build.

danpetry replied:
done

JeanChristopheMorinPerso commented on recipe/build.sh:
```diff
@@ -48,7 +56,14 @@ ${PYTHON} tools/ci_build/build.py \
 
-cp build-ci/Release/dist/onnxruntime-*.whl onnxruntime-${PKG_VERSION}-py3-none-any.whl
-${PYTHON} -m pip install onnxruntime-${PKG_VERSION}-py3-none-any.whl
+cp build-ci/Release/dist/${WHL_BASE_NAME}-*.whl ${WHL_BASE_NAME}-${PKG_VERSION}-py3-none-any.whl
```
Same here, the copy isn't necessary.

danpetry replied:
done

danpetry commented:
Thanks, due to the long build times I'll wait until I get another review before making changes.

Are these the only changes you want before an approval? if not, would be cool to get any other changes before I implement these, for the same reason, if you will be making more comments that is.

danpetry commented:
CUDA build takes 24 hours

JeanChristopheMorinPerso reviewed (approved):
LGTM, on the condition that the comments are addressed before merging.

cbouss reviewed (approved):
LGTM, there is a lot of vendored stuff in the logs but that can be managed later.

JeanChristopheMorinPerso commented:
>LGTM, there is a lot of vendored stuff in the logs but that can be managed later.

I saw that too and came to the same conclusion as you.

danpetry commented:
Re protobuf, at least, I decided not to: https://github.com/conda-forge/onnxruntime-feedstock/issues/57#issuecomment-1518033552

danpetry committed:
Remove useless copy of wheels

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_42xe8xyupk/clone` failed. (See above for error)
The following problems have been found:

===== WARNINGS =====
- clone/recipe/meta.yaml:43: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:53: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:45: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:54: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:56: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
===== Final Report: =====
0 Errors and 5 Warnings were found

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_8cqxowj8kc/clone` failed. (See above for error)
The following problems have been found:

===== WARNINGS =====
- clone/recipe/meta.yaml:54: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:53: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:45: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:43: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:56: host_section_needs_exact_pinnings: Linked libraries host should have exact version pinnings.
===== Final Report: =====
0 Errors and 5 Warnings were found

</details>
